File: apple/Frontrow/Models/TeamStats.swift
import Foundation

/// Derived team numbers — the Swift half of the web's `team-stats.ts`, so the
/// app and the website agree on what a streak or a win percentage is.
enum TeamStats {
    struct Record: Sendable {
        let wins: Int, losses: Int, ties: Int
    }

    /// "84-62" or "10-4-1" → the parts, or nil if it isn't a record.
    static func parseRecord(_ text: String) -> Record? {
        let parts = text.split(separator: "-").map { Int($0.trimmingCharacters(in: .whitespaces)) }
        guard parts.count >= 2, let w = parts[0], let l = parts[1] else { return nil }
        let t = parts.count > 2 ? (parts[2] ?? 0) : 0
        return Record(wins: w, losses: l, ties: t)
    }

    /// ".597" — three decimals, no leading zero, as the standings print it.
    static func winPct(_ r: Record) -> String {
        let games = r.wins + r.losses + r.ties
        guard games > 0 else { return "—" }
        let pct = (Double(r.wins) + Double(r.ties) / 2) / Double(games)
        let text = String(format: "%.3f", pct)
        return text.hasPrefix("0") ? String(text.dropFirst()) : text
    }

    struct Streak: Sendable {
        let label: String
        let result: Outcome
    }

    /// "W3" from the form list, which arrives most-recent first.
    static func currentStreak(_ form: [FormEntry]) -> Streak? {
        guard let first = form.first else { return nil }
        let run = form.prefix { $0.result == first.result }.count
        return Streak(label: "\(first.result.rawValue)\(run)", result: first.result)
    }

    static func mean(_ xs: [Int]) -> Double {
        xs.isEmpty ? 0 : Double(xs.reduce(0, +)) / Double(xs.count)
    }
}

File: src/lib/team-stats.ts
/* Derived team stats shared by the dashboard's season-stat cards and the
   iOS app's Teams tab: record parsing, win %, streak and per-game means.
   Pure functions over the TeamCard shape — no fetching, no React. */

import type { LeagueId, Outcome } from "@/lib/types";

/** Noun for the scoring unit in a league, for labels like "Runs/G". */
export const SCORE_NOUN: Record<string, string> = {
  mlb: "Runs",
  nhl: "Goals",
  nfl: "Points",
  nba: "Points",
  "college-football": "Points",
};

export function scoreNoun(league: LeagueId): string {
  return SCORE_NOUN[league] ?? "Points";
}

export function parseRecord(rec: string): { w: number; l: number; t: number } | null {
  const parts = rec.split("-").map((n) => parseInt(n.trim(), 10));
  if (parts.length < 2 || Number.isNaN(parts[0]) || Number.isNaN(parts[1])) return null;
  return { w: parts[0], l: parts[1], t: Number.isNaN(parts[2]) ? 0 : parts[2] ?? 0 };
}

/** ".597" — win percentage without a leading zero. */
export function winPct(w: number, l: number, t: number): string {
  const games = w + l + t;
  if (games === 0) return "—";
  return ((w + t / 2) / games).toFixed(3).replace(/^0/, "");
}

/** Current streak from the form list (most-recent first), e.g. "W3". */
export function currentStreak(form: { result: Outcome }[]): { label: string; result: Outcome } | null {
  if (!form.length) return null;
  const result = form[0].result;
  let n = 0;
  for (const g of form) {
    if (g.result === result) n++;
    else break;
  }
  return { label: `${result}${n}`, result };
}

export const mean = (xs: number[]) => (xs.length ? xs.reduce((a, b) => a + b, 0) / xs.length : 0);
